### DISCUSSION TITLE
Detailed Forecast pv total <> Forecast Today

### DISCUSSION THREAD

**Discussion by Ashleysbox** [BJReplay/ha-solcast-solar#168] [Q&A] [answered]
Hi All,

One of my sensors totals the Half hourly PV Estimate over a user definable time period for sensor.solcast_pv_forecast_forecast_today however the result I'm getting does not match the overall Solar Forecast Today 

For example sensor.solcast_pv_forecast_forecast_today = 5.86 for me today but the sum of the half hourly detail pv_estimates = 11.7

I'm using this code to sum the pv_estimates

```
{% set forecast_today = state_attr('sensor.solcast_pv_forecast_forecast_today', 'detailedForecast')  %}
{% set ns = namespace(additional = 0) %}
{% for record in forecast_today if ( record.period_start < today_at("23:30") ) %}
      {% set ns.additional = ns.additional + record.pv_estimate %}
{% endfor %}
{{ns.additional}}

```
Anyone got any suggestions as to why the detailed pv total and headline forecast don't match please?

(Cut and paste of the detailed data which confirms the total is 11.7 below)

Dayname
Tuesday
DataCorrect
true
DetailedForecast
- period_start: '2024-09-10T00:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T00:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T01:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T01:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T02:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T02:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T03:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T03:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T04:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T04:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T05:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T05:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T06:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T06:30:00+01:00'
pv_estimate: 0.0042
pv_estimate10: 0.002
pv_estimate90: 0.0062
- period_start: '2024-09-10T07:00:00+01:00'
pv_estimate: 0.0334
pv_estimate10: 0.023
pv_estimate90: 0.048
- period_start: '2024-09-10T07:30:00+01:00'
pv_estimate: 0.0792
pv_estimate10: 0.0542
pv_estimate90: 0.1146
- period_start: '2024-09-10T08:00:00+01:00'
pv_estimate: 0.1524
pv_estimate10: 0.1
pv_estimate90: 0.2506
- period_start: '2024-09-10T08:30:00+01:00'
pv_estimate: 0.2724
pv_estimate10: 0.156
pv_estimate90: 0.4128
- period_start: '2024-09-10T09:00:00+01:00'
pv_estimate: 0.4092
pv_estimate10: 0.247
pv_estimate90: 0.6382
- period_start: '2024-09-10T09:30:00+01:00'
pv_estimate: 0.5624
pv_estimate10: 0.3518
pv_estimate90: 0.8652
- period_start: '2024-09-10T10:00:00+01:00'
pv_estimate: 0.4176
pv_estimate10: 0.2308
pv_estimate90: 0.6713
- period_start: '2024-09-10T10:30:00+01:00'
pv_estimate: 0.6877
pv_estimate10: 0.4128
pv_estimate90: 1.0554
- period_start: '2024-09-10T11:00:00+01:00'
pv_estimate: 0.8835
pv_estimate10: 0.5664
pv_estimate90: 1.2019
- period_start: '2024-09-10T11:30:00+01:00'
pv_estimate: 0.8242
pv_estimate10: 0.4568
pv_estimate90: 1.1916
- period_start: '2024-09-10T12:00:00+01:00'
pv_estimate: 0.9554
pv_estimate10: 0.5876
pv_estimate90: 1.278
- period_start: '2024-09-10T12:30:00+01:00'
pv_estimate: 1.0496
pv_estimate10: 0.6296
pv_estimate90: 1.3459
- period_start: '2024-09-10T13:00:00+01:00'
pv_estimate: 1.0152
pv_estimate10: 0.5806
pv_estimate90: 1.332
- period_start: '2024-09-10T13:30:00+01:00'
pv_estimate: 0.9178
pv_estimate10: 0.4628
pv_estimate90: 1.3032
- period_start: '2024-09-10T14:00:00+01:00'
pv_estimate: 0.8178
pv_estimate10: 0.3554
pv_estimate90: 1.2456
- period_start: '2024-09-10T14:30:00+01:00'
pv_estimate: 0.717
pv_estimate10: 0.2616
pv_estimate90: 1.1675
- period_start: '2024-09-10T15:00:00+01:00'
pv_estimate: 0.6046
pv_estimate10: 0.1706
pv_estimate90: 1.1195
- period_start: '2024-09-10T15:30:00+01:00'
pv_estimate: 0.4868
pv_estimate10: 0.1052
pv_estimate90: 1.0295
- period_start: '2024-09-10T16:00:00+01:00'
pv_estimate: 0.343
pv_estimate10: 0.063
pv_estimate90: 0.8449
- period_start: '2024-09-10T16:30:00+01:00'
pv_estimate: 0.205
pv_estimate10: 0.0336
pv_estimate90: 0.6122
- period_start: '2024-09-10T17:00:00+01:00'
pv_estimate: 0.1156
pv_estimate10: 0.019
pv_estimate90: 0.4224
- period_start: '2024-09-10T17:30:00+01:00'
pv_estimate: 0.0758
pv_estimate10: 0.0148
pv_estimate90: 0.2814
- period_start: '2024-09-10T18:00:00+01:00'
pv_estimate: 0.0484
pv_estimate10: 0.0106
pv_estimate90: 0.161
- period_start: '2024-09-10T18:30:00+01:00'
pv_estimate: 0.0276
pv_estimate10: 0.0064
pv_estimate90: 0.0912
- period_start: '2024-09-10T19:00:00+01:00'
pv_estimate: 0.0084
pv_estimate10: 0.0022
pv_estimate90: 0.0298
- period_start: '2024-09-10T19:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T20:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T20:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T21:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T21:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T22:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T22:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T23:00:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
- period_start: '2024-09-10T23:30:00+01:00'
pv_estimate: 0
pv_estimate10: 0
pv_estimate90: 0
F6ee-b0d1-26e2-216b
2.93
Estimate90-f6ee-b0d1-26e2-216b
4.53
3f48-e585-5653-3839
2.93
Estimate90-3f48-e585-5653-3839
4.83
Estimate90
9.36

**Comment by autoSteve** [comment, marked as answer]
Divide your half hourly detail pv_estimates by two. The values are average kW for the half hour period. This is not a kWh (KiloWatt hour), so simply divide the total to get an estimate energy for the hour.

Or use the detailed hourly attribute, which is what it does...

**Reply by Ashleysbox** [reply]
Many thanks - that's what I had missed

**Reply by gcoan** [reply]
> Divide your half hourly detail pv_estimates by two. The values are average kW for the half hour period. This is not a kWh (KiloWatt hour),

@autoSteve might be worth explicitly highlighting this in the readme

**Reply by autoSteve** [reply]
Maybe. We do cover it in discussions, and frankly template sensor usage of attributes like this are beyond 101 / readme.

But maybe the readme needs a 301 section... 🫣 Maybe a 201/301.md added... which nobody would read. 🤔

I'm happy answering stuff in discussions. It's really infrequent, and keeps me on my toes.

**Reply by autoSteve** [reply]
And everybody is smarter now because discussion, so it's a win-win. I don't have to update docs, folks read this. Leaving this one open for posterity.

**Comment by autoSteve** [comment]
(11.7 / 2 = 5.85.)

Or use the custom "X hour" sensor instead of yours. That one utilises spline interpolated values for part-way through a half-hour period, with the restriction that splines are calculated for the current 24-hour period. (Beyond the 24-hour period it uses linear interpolation, which is good enough.) So if your period is say eight hours it'll do a better job of 'now' to then than you could.

**Reply by Ashleysbox** [reply]
I'll take a look,  I should be able to use it instead

**Comment by autoSteve** [comment]
(Given the whole "user selectable" thing, this begs the question in my mind, "should the custom X hour sensor have as attributes hours from one to twenty four as values?" Then for a template sensor, grab the number of hours required from the attribute. The only downside I see is slight increased processor usage, but that would be insignificant, given splines are calculated when forecasts are retrieved, which is the heavy lifting, and sensor update every five minutes is just a couple of dictionary lookups unless beyond spline calc period, which is no mean math.)
